ccmbioinfo/stager
# Flask: implement email notifications

- We used SendGrid Email API to send emails to appropriate users when an analysis is requested.
- The `send_email` function will send from whichever user email we specify in the `from_email` field. The email domains in `from_email` would need to be verified for some email providers (like SickKids) to accept it and show it in the end user's inbox. So far, we've verified the `ccmdev.ca` domain, and the test `from_email` is `test@ccmdev.ca`.
- In the email, there would also be an Unsubscribe option, where users can choose to not receive future notification emails. 
- The code also checks for whether we have exceeded the email limit in the day, and if so, queue the email the next day. Otherwise, it sends the email 1 minute after an analysis has been successfully requested.

**To test the code, you need to:**

1. Create a SendGrid account and an API key and add it to `SENDGRID_API_KEY` in `.env`.
2. Follow these steps in [Single Sender Verification](https://docs.sendgrid.com/ui/sending-email/sender-verification) to verify yourself as a sender. If you are not verified as a sender, you will not be able to send any email.Hence, the recipient (which is also you in this case, because you're testing) won't receive any email. 
_Note:_ Single Verification is only used for fast, local testing and not recommended in production. As mentioned above, we are going to verify the email sender domains, similar to `ccmdev.ca`, in production.
3. Add the email you have verified in step 2 to the `from_email` argument in `email.send_email`
4. Add your personal (non-SickKids) email to the `to` field. The reason why you can't use your SickKids email here is because SickKids email server doesn't trust emails sent via sendgrid as a proxy. This problem is solved once we verify the domain. 
5. Import [this](https://drive.google.com/file/d/1qgtVTJbpqtqpjSJoUOlybzjBemeyXHzp/view?usp=sharing) custom email HTML template to the **[Dynamic Templates](https://mc.sendgrid.com/dynamic-templates)** section in your SendGrid dashboard.
6. Copy-paste the template ID and add it to `.env`.
7. Create an Unsubscribe group in your dashboard [here](https://app.sendgrid.com/suppressions/advanced_suppression_manager). Add your unsubscribe group ID to `.env`.
5. Try requesting an analysis and you should see an email to your personal inbox!

**Note:**

This code can be merged into master with email notification disabled by simplying setting the `SENDGRID_API_KEY` to empty until we've properly confirmed all collaborators that we will be sending them email.
